Public/Src/Engine/Scheduler/PipExecutor.cs

```diff
@@ -934,6 +934,40 @@ private static void MakeSharedOpaqueOutputIfNeeded(IPipExecutionEnvironment envi
             }
         }
 
+        /// <summary>
+        /// Report change affected ouputs to file content manager
+        /// </summary>
+        internal static void ReportSourceChangeAffectedOutputs(
+            IPipExecutionEnvironment environment,
+            PipResultStatus status,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedFiles,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedEnumerations,
+            Pip pip,
+            IReadOnlyCollection<(FileArtifact, FileMaterializationInfo, PipOutputOrigin)> outputContent = null,
+            IReadOnlyCollection<DirectoryArtifact> directoryOutputs = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputFiles = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputDirectroies = null)
+        {
+            PipOutputOrigin? overrideOutputOrigin = null;
+            if (status == PipResultStatus.NotMaterialized)
+            {
+                overrideOutputOrigin = PipOutputOrigin.NotMaterialized;
+            }
+
+            if (pip.IsOutputAffectedBySourceChange(dynamicallyObservedFiles, dynamicallyObservedEnumerations, environment.Context.PathTable, sourceChangeAffectedOutputFiles, sourceChangeAffectedOutputDirectroies))
+            {
+                foreach (var output in outputContent ?? ReadOnlyArray<(FileArtifact, FileMaterializationInfo, PipOutputOrigin)>.Empty)
+                {
+                    environment.State.FileContentManager.ReportSourceChangeAffectedOutputs(output.Item1, overrideOutputOrigin ?? output.Item3, true);

```

Comment:
>environment.State.FileContentManager [](start = 20, length = 36)

Why doe this reporting have to go through `FileContentManager`? `PipExecutor` is basically our scheduler, so can we not report directly to the scheduler? Also, the `FileContentManager` does not hold any state regarding this report.

Reply:
moved to IPipExecutionEnvironment.